xdelta3-dir-patcher

```diff
@@ -196,35 +665,69 @@ class XDelta3DirPatcher(object):
-
-            print("Leaving directory", root)
-
-        print("Cleaning up...")
-        rmtree(patch_dir)
+        print('Enumerating old files')
+        old_files = []
+        with XDeltaArchive(old_dir) as old_archive:
+            old_files = old_archive.list_files()
+        if self.args.verbose: print("Old: %s" % old_files)
+
+        print('Applying patches from %s' % patch_bundle)
+        with XDeltaArchive(patch_bundle) as patch_archive:
+            all_archive_items = patch_archive.list_files()
+            if self.args.verbose: print("All in patch: %s" % all_archive_items)
+
+            patches = [p for p in all_archive_items if p.startswith(delta_patch_root)]
+            if self.args.verbose: print("Patches: %s" % patches)
+
+            # XXX: We need to strip out the internal path prefix on the patches
+            #      to be able to compare the file lists and build a "to_remove"
+            #      array. The reason why we use len() vs path.relpath() is so that
+            #      we don't strip out the trailing path.sep() on directories that
+            #      relpath does automatically
+            files_in_patch = []
+            for filename in patches:
+                relative_filename = filename[len(delta_patch_root) + len(path.sep):]
+                files_in_patch.append(relative_filename)
+
+            if self.args.verbose: print("In patch: %s" % files_in_patch)
+
+            removed_items = [f for f in old_files if f not in files_in_patch]
+
+            # XXX: This is to ensure that we never try to remove a directory while
+            #      all of its children are in queue waiting for a thread (causing
+            #      a deadlock)
+            removed_items.sort(key=len, reverse=True)
+
+            if self.args.verbose: print("Removed: %s" % removed_items)
+
+            print("Removing deleted files")
+
+            for removed_item in removed_items:
+                if self.args.debug:
+                    print('Queueing deletion %s' % filename)
+                else:
+                    print('x', end = "")
+
+                runner.add_task(self.remove_item, (target_dir,
+                                                   removed_item,
+                                                   self.args.debug))
+
+            for patch in patches:
+                if self.args.debug:
+                    print('Queueing %s' % patch)
+                else:
+                    print('.', end = "")
+
+                runner.add_task(self._apply_file_delta, (patch_archive,
+                                                         patch,
+                                                         old_dir,
+                                                         target_dir,
+                                                         delta_patch_root,
+                                                         patch_staging_dir))
+            runner.join_all()
```

Comment:
I think it would be safer to wait for the deletion tasks prior to the patching tasks. I'm worried about what's going to happen with the directories with these threads running concurrently.